### ISSUE TITLE
sp_BlitzCache - Bug in duplicate plan query

### ISSUE THREAD

**Issue by erikdarlingdata** [BrentOzarULTD/SQL-Server-First-Responder-Kit#2909]
**Version of the script**
`@Version = '8.03', @VersionDate = '20210420'`

**What is the current behavior?**
There's a bug in the section of code that starts with `Checking for single use plans and plans with many queries`

Specifically, in this section, we're getting a count of distinct plan handles grouped by query hash, which makes no sense. I believe this is why the >100% numbers were coming back.

```
    SELECT SUM(x.duplicate_plan_handles) AS duplicate_plan_handles
    FROM (
        SELECT COUNT_BIG(DISTINCT plan_handle) AS duplicate_plan_handles
        FROM sys.dm_exec_query_stats qs
            CROSS APPLY sys.dm_exec_plan_attributes(qs.plan_handle) pa
        WHERE pa.attribute = N'dbid'
        GROUP BY qs.query_hash, pa.value
        HAVING COUNT_BIG(DISTINCT plan_handle) > 5
    ) AS x
```

It may also make sense to modify this CTE to get a count from `dm_exec_query_stats` instead of `dm_exec_cached_plans`:

```
    SELECT COUNT_BIG(*) AS total_plans
    FROM sys.dm_exec_cached_plans AS deqs
    WHERE deqs.cacheobjtype = N'Compiled Plan'
```

Since that would match the section of [code immediately above it](https://github.com/BrentOzarULTD/SQL-Server-First-Responder-Kit/blob/dev/sp_BlitzCache.sql#L1582) that looks for plan cache instability. 

**If the current behavior is a bug, please provide the steps to reproduce.**
Code review

**What is the expected behavior?**
Be logically correct

**Which versions of SQL Server and which OS are affected by this issue? Did this work in previous versions of our procedures?**
All.

Just a note, I'm going to leave this open for a few days for feedback. Not sure if I'm missing something obvious, here.

**Comment by erikdarlingdata** [comment]
Working through this a little and talking it over with Smart People™, it likely makes more sense to compare query plan hashes, since plan handles could result from multiple sources legitimately executing the same query.

There's also some oddballness between the datatypes that the query generates and the datatypes defined in the temp table they get inserted to.

**Comment by BrentOzar** [comment]
Ooo, this is so tricky. I took the repro query for #2900 and modified it:

```
/* One time setup */
USE StackOverflow;
GO
CREATE INDEX Location ON dbo.Users(Location);
GO
CREATE OR ALTER PROC dbo.usp_TestIndia AS
SELECT TOP 101 * FROM dbo.Users WHERE Location = N'India' ORDER BY Reputation DESC;
GO
CREATE OR ALTER PROC dbo.usp_TestUSA AS
SELECT TOP 101 * FROM dbo.Users WHERE Location = N'USA' ORDER BY Reputation DESC;
GO
CREATE OR ALTER PROC dbo.usp_TestCanada AS
SELECT TOP 101 * FROM dbo.Users WHERE Location = N'Canada' ORDER BY Reputation DESC;
GO


/* Run this part for testing: */
DROP TABLE IF EXISTS #Locations;
SELECT TOP 20 Location
	INTO #Locations
	FROM dbo.Users
	WHERE Location <> ''
	GROUP BY Location
	ORDER BY COUNT(*) DESC;
GO
DBCC FREEPROCCACHE;
GO
EXEC usp_TestIndia;
EXEC usp_TestUSA;
EXEC usp_TestCanada;
GO
DECLARE @Location NVARCHAR(100), @StringToExec NVARCHAR(4000);
DECLARE curses CURSOR FOR SELECT Location FROM #Locations;
OPEN curses;
FETCH NEXT FROM curses INTO @Location
WHILE @@FETCH_STATUS = 0
BEGIN
	SET @StringToExec = N'SELECT TOP 101 * FROM dbo.Users WHERE Location = N''' + @Location + N''' ORDER BY Reputation DESC;'
	EXEC(@StringToExec);
	FETCH NEXT FROM curses INTO @Location;
END

CLOSE curses;
DEALLOCATE curses;
GO

EXEC sp_BlitzCache
GO
```

The current release of sp_BlitzCache does this:
<img width="1042" alt="Screen Shot 2021-05-20 at 6 36 25 AM" src="https://user-images.githubusercontent.com/245462/118930844-c5ed4a80-b935-11eb-8fe1-7c4047c2d2fc.png">

* Warnings - Multiple Plans (23) - that's the code from #2900, and that's a problem, but I'll hit that over in #2900. The problem is that the stored procs (which are different code) are also being included in the Multiple Plans number. (sigh) That issue is going to be a pain the more that I look at it.
* Priority 1 - Many Duplicate Plans - 96% are duplicates - this is wrong because it's including the stored procs.
* Priority 1 - Many Single-Use Plans - 96% are single use - this is correct. (The procs have only been run once.)

The pull request for this issue shows this:
<img width="986" alt="Screen Shot 2021-05-20 at 6 40 11 AM" src="https://user-images.githubusercontent.com/245462/118931218-3eeca200-b936-11eb-9e24-33ccae070a98.png">
* Warnings - Multiple Plans (10) - that's okay, doesn't matter, working on that over in #2900.
* Priority 1 - Many Duplicate Plans - this is a problem because it doesn't show up even though we have duplicate plans.
* Priority 1 - Many Single-Use Plans - 81% - this is wrong.

We do have a problem that we need to fix though, heh, so leaving this open so we can figure out the right fix. I really wanna work on this but I'm tight on time right now.

**Comment by erikdarlingdata** [comment]
This is doable! 

We'd just have to left join to `dm_exec_procedure_stats` on `sql_handle` to get the parent `object_id` to differentiate.

**Comment by erikdarlingdata** [comment]
Alright, I think I have this working. 

Here's what I get using your Adonis-like demo code. To help validate results, I'm using the `many_plans` portion of the CTE chain, but I'm including `query_hash` and `query_plan_hash` in the output, and I'm only restricting the count to `> 1` rather than `> 5`. 

* All queries have a `query_hash` of `0x2092C88D26FF4523`
* `0xD135EC511C700301` has a count of 13, which lines up with the individual statements from the cursor, even though it also shows up as the hash for `TestUsa`
* `0x7BE9D4A5E6386376` has a count of 4 from the cursor, and a 5th entry that's excluded because it's from `TestCanada`
* `0xF66758059DD14E60` has a count of 3 from the cursor, and a 4th entry that's excluded because it's from `TestIndia`

![image](https://user-images.githubusercontent.com/2136037/119229518-68e4c680-bae6-11eb-8c6e-480fb9d7a451.png)

**Comment by BrentOzar** [comment]
Merging this in for the "Many Duplicate Plans" warning fix, and then I'll work on the "Multiple Plans (X)" over in the other issue. Thanks again for this, man! You rock.

**Comment by BrentOzar** [comment]
DOH! I'm an idiot, the "Multiple Plans (X)" is already completely fixed in #2900. Merging both of these gives us one complete picture. (sigh)
